### PR TITLE
Drive IO stats

### PR DESCRIPTION
**Pull request by r-scheele**
Implemented new metrics in `publishDriveStats`:
   - Online/Offline status of drives.
   - Total bytes read and written.
   - Read and write latency calculations.
   - Read and write throughput.
   - Drive wait time.
Similar to the existing implementation `publishVolumeStats`

### CODE REVIEW

File: pkg/metrics/collector.go [hunk 1]
```diff
@@ -96,21 +102,254 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+			"unable to find device by FSUUID %v; "+
+				"either device is removed or run command "+
+				"`sudo udevadm control --reload-rules && sudo udevadm trigger`"+
+				" on the host to reload", drive.Status.FSUUID)
+
+		return
+	}
+
+	// Online/Offline Status
+	if _, err := os.Stat("/sys/block/" + device); os.IsNotExist(err) {
```
Comment (Praveenrajmani) [inline pkg/metrics/collector.go:134]: This will not work for partitions. The partitions info will be under `/sys/block/<parent-dev>/<dev>/`

File: pkg/metrics/collector.go [hunk 2]
```diff
@@ -96,21 +102,254 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+	}
+	stats.TimeInQueue, err = strconv.ParseInt(fields[9], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing TimeInQueue: %w", err)
+	}
+
+	return &stats, nil
+}
+
+func getSectorSize(device string) int64 {
```
Comment (Praveenrajmani) [inline pkg/metrics/collector.go:291]: let this function return an (error, int64)

File: pkg/metrics/collector.go [hunk 3]
```diff
@@ -96,21 +102,254 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+	ReadTicks    int64
+	WriteSectors int64
+	WriteTicks   int64
+	TimeInQueue  int64
+}
+
+func (c *metricsCollector) publishDriveStats(ctx context.Context, drive *types.Drive, ch chan<- prometheus.Metric) {
+	device, err := c.getDeviceByFSUUID(drive.Status.FSUUID)
+	if err != nil {
+		klog.ErrorS(
```
Comment (Praveenrajmani) [inline pkg/metrics/collector.go:116]: If the device is not found by its FSUUID, this means the drive is not READY. we should also export this information as a metric.

File: pkg/metrics/collector.go [hunk 4]
```diff
@@ -96,21 +102,254 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+
+		return
+	}
+
+	// Online/Offline Status
+	if _, err := os.Stat("/sys/block/" + device); os.IsNotExist(err) {
+		// Drive is offline
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				prometheus.BuildFQName(consts.AppName, "stats", "drive_status"),
```
Comment (Praveenrajmani) [inline pkg/metrics/collector.go:138]: can we have `_drive_ready` instead of `_drive_status`? 

if these directories are not present, then it means the drive is "not ok" (or) "not ready". we can have a label name accordingly.

**Review by Praveenrajmani — COMMENTED** [review]
can you fix the lint errors? @r-scheele

File: pkg/metrics/collector.go (hunk 1 above)
Comment (r-scheele) [reply pkg/metrics/collector.go:134]: `device, err := c.getDeviceByFSUUID(drive.Status.FSUUID)` returns the device in this format `/dev/dm-1`, so there's no way access the partitions. in my case `dm-0` and `dm-1` are both drives.

File: pkg/metrics/collector.go [hunk 5]
```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+	if err != nil {
+		return nil, fmt.Errorf("parsing TimeInQueue: %w", err)
+	}
+
+	return &stats, nil
+}
+
+func getSectorSize(device string) (int64, error) {
+	// Construct the file path to the sector size file
+	filePath := "/sys/block/" + device + "/queue/hw_sector_size"
```
Comment (balamurugana) [inline pkg/metrics/collector.go:307]: 1. This file doesn't exist for disk partitions. For partitions, find it is disk and fetch sector size from it and use it.
2. if `/sys/class/block/<device>/queue/hw_sector_size` is missing for disk, use 512 as default block size.
3. Move this function as `GetHardwareSectorSize` to `pkg/device/sysfs_linux.go`

File: pkg/metrics/collector.go [hunk 6]
```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
-	resultCh := volume.NewLister().
+	// Collecting volume statistics
+	volumeResultCh := volume.NewLister().
 		NodeSelector([]directpvtypes.LabelValue{directpvtypes.ToLabelValue(string(c.nodeID))}).
 		List(ctx)
-	for result := range resultCh {
+	for result := range volumeResultCh {
 		if result.Err != nil {
-			return
+			continue
```
Comment (balamurugana) [inline pkg/metrics/collector.go:348]: Why `continue`? do you mean `break`?

File: pkg/metrics/collector.go [hunk 7]
```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
 		}
 	}
+
+	// Collecting drive statistics
+	driveResultCh := drive.NewLister().
+		NodeSelector([]directpvtypes.LabelValue{directpvtypes.ToLabelValue(string(c.nodeID))}).
+		List(ctx)
+	for result := range driveResultCh {
+		if result.Err != nil {
+			continue
```
Comment (balamurugana) [inline pkg/metrics/collector.go:362]: Why `continue`? do you mean `break`?

**Review by balamurugana — CHANGES_REQUESTED** [review]
Please add respective functional tests

File: pkg/metrics/collector.go (hunk 7 above)
Comment (r-scheele) [reply pkg/metrics/collector.go:362]: I used `continue` to ensure we process all drives, even if an error occurs on one of them.

File: pkg/metrics/collector.go (hunk 5 above)
Comment (r-scheele) [reply pkg/metrics/collector.go:307]: I found a `defaultBlockSize` as 512 in the `sysfs_linux.go`, Since it's the same value, should I go ahead and use it? but I'm thinking block size and sector size aren't the same thing, I could create a new constant just for sector size.

File: pkg/metrics/collector.go [hunk 8]
```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+func (c *metricsCollector) publishDriveStats(drive *types.Drive, ch chan<- prometheus.Metric) {
+	device, err := c.getDeviceByFSUUID(drive.Status.FSUUID)
+	if err != nil {
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				prometheus.BuildFQName(consts.AppName, "stats", "drive_ready"),
+				"Drive Online/Offline Status",
+				[]string{"drive"}, nil),
+			prometheus.GaugeValue,
+			0, // 0 for offline
```
[the remark this answers is not in the document]
Comment (r-scheele) [reply pkg/metrics/collector.go:122]: If I get you correctly, I'm using `0` to show offline and `1` to show online. As it seems more Boolean-like. Prometheus does not support  a native Boolean data type.

File: .gitignore [hunk 9]
```diff
@@ -11,3 +11,4 @@ dist/
 kustomize
 operator-sdk
 kubectl-directpv_*
+directpv.tar
```
[the remark this answers is not in the document]
Comment (r-scheele) [reply .gitignore:14]: it was added to the `.gitignore` because it is generated during testing. 
https://github.com/minio/directpv/blob/master/docs/developer.md the image tar ball was created by podman

**Comment by r-scheele** [comment]
> Please add respective functional tests



> Please add respective functional tests

Yes i will, I'm going through the previous one written already.

File: .gitignore (hunk 9 above)
[the remark this answers is not in the document]
Comment (balamurugana) [reply .gitignore:14]: Add a new line

File: pkg/metrics/collector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/metrics/collector.go:122]: What is the other way to do to indicate boolean state? If none, default constants for drive online/offline and user it here.

File: pkg/metrics/collector.go (hunk 5 above)
Comment (balamurugana) [reply pkg/metrics/collector.go:307]: Most of the system have default block/sector size as 512.  you could define default sector size as default block size.

File: pkg/metrics/collector.go (hunk 7 above)
Comment (balamurugana) [reply pkg/metrics/collector.go:362]: Once the error is thrown, we cannot get values further. `break` is logically correct thing to do.

File: pkg/metrics/collector.go (hunk 1 above)
Comment (Praveenrajmani) [reply pkg/metrics/collector.go]: Ignore this.. `/sys/block/<dev>/stat` will work. There are no any special cases for partitions. I see we have symlinks appropriately.

**Comment by Praveenrajmani** [comment]
PTAL @r-scheele

**Comment by r-scheele** [comment]
> PTAL @r-scheele

Hi @Praveenrajmani yes, i'm working on this review now

**Comment by r-scheele** [comment]
Hi @balamurugana  @Praveenrajmani PTAL?

File: pkg/admin/installer/directpv.min.io_directpvdrives.yaml [hunk 10]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.15.0
+    controller-gen.kubebuilder.io/version: v0.14.0
```
Comment (Praveenrajmani) [inline pkg/admin/installer/directpv.min.io_directpvdrives.yaml:6]: please update your controller-gen version and run `build.sh` again

File: pkg/installer/directpv.min.io_directpvvolumes.yaml [hunk 11]
```diff
@@ -0,0 +1,149 @@
+---
+apiVersion: apiextensions.k8s.io/v1
```
Comment (Praveenrajmani) [inline pkg/installer/directpv.min.io_directpvvolumes.yaml:2]: this file looks unrelated, can you remove?

File: cmd/kubectl-directpv/discover.go [hunk 12]
```diff
@@ -218,7 +218,7 @@ func discoverMain(ctx context.Context) {
 				eprintf(true, "discovery failed; %v\n", err)
 				os.Exit(1)
 			case <-ctx.Done():
-				eprintf(true, ctx.Err().Error())
+				eprintf(true, "%v", ctx.Err())
```
Comment (Praveenrajmani) [inline cmd/kubectl-directpv/discover.go:221]: why are we changing this?

**Review by Praveenrajmani — COMMENTED** [review]
Initial review

File: cmd/kubectl-directpv/discover.go (hunk 12 above)
Comment (harshavardhana) [reply cmd/kubectl-directpv/discover.go:221]: eprintf() requires `bool, format, data` I think that's why  and the new Go vet complains for incorrect usage.

File: pkg/admin/installer/directpv.min.io_directpvdrives.yaml (hunk 10 above)
Comment (r-scheele) [reply pkg/admin/installer/directpv.min.io_directpvdrives.yaml:6]: Okay, i did that already.

File: pkg/installer/directpv.min.io_directpvvolumes.yaml (hunk 11 above)
Comment (r-scheele) [reply pkg/installer/directpv.min.io_directpvvolumes.yaml:2]: it was present in the file before https://github.com/minio/directpv/blob/master/pkg/admin/installer/directpv.min.io_directpvvolumes.yaml

Comment (Praveenrajmani) [reply pkg/installer/directpv.min.io_directpvvolumes.yaml:2]: these files are not present in the existing source. can you remove these?

File: pkg/device/sysfs_linux.go [hunk 13]
```diff
@@ -106,3 +109,95 @@ func getHolders(name string) ([]string, error) {
...
+		return nil, driveStatus, fmt.Errorf("no valid statistics found in file %s", filePath)
+	}
+
+	return stats, driveStatus, nil
+}
+
+// GetHardwareSectorSize retrieves the hardware sector size for a given device.
+// It works for both whole disks and partitions.
+// The device name should be without the "/dev/" prefix (e.g., "sda" or "sda1").
+func GetHardwareSectorSize(deviceName string) (uint64, error) {
```
Comment (Praveenrajmani) [inline pkg/device/sysfs_linux.go:166]:
```suggestion
func HardwareSectorSize(deviceName string) (uint64, error) {
```

File: pkg/device/sysfs_linux.go [hunk 14]
```diff
@@ -106,3 +109,95 @@ func getHolders(name string) ([]string, error) {
...
+		return 0, fmt.Errorf("failed to read hardware sector size for %s: %v", deviceName, err)
+	}
+
+	// Parse the sector size as an integer
+	sectorSize, err := strconv.ParseUint(strings.TrimSpace(string(content)), 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse hardware sector size for %s: %v", deviceName, err)
+	}
+
+	klog.Infof("Hardware sector size for %s: %d bytes", deviceName, sectorSize)
```
Comment (Praveenrajmani) [inline pkg/device/sysfs_linux.go:201]: Same: Use verbose levels in logging

**Comment by Praveenrajmani** [comment]
sorry for the late review @r-scheele

File: pkg/metrics/collector.go [hunk 15]
```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
...
+			"unable to find device by FSUUID %v; "+
+				"either device is removed or run command "+
+				"`sudo udevadm control --reload-rules && sudo udevadm trigger`"+
+				" on the host to reload", drive.Status.FSUUID)
+
+		return
+	}
+	device = strings.TrimPrefix(device, "/dev")
+	// Online/Offline Status
+	if _, err := os.Stat("/sys/block" + device); os.IsNotExist(err) {
```
[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply pkg/metrics/collector.go]: ^^ @r-scheele

**Comment by r-scheele** [comment]
> sorry for the late review @r-scheele

No problem @Praveenrajmani, would look into it soon.

File: pkg/device/sysfs_linux.go [hunk 16]
```diff
@@ -106,3 +109,95 @@ func getHolders(name string) ([]string, error) {
 func getDMName(name string) (string, error) {
 	return readFirstLine("/sys/class/block/" + name + "/dm/name")
 }
+
+// GetStat retrieves and returns statistics for a given device name.
+func GetStat(name string) ([]uint64, int, error) {
+	filePath := path.Join("/sys/class/block/", name, "/stat")
+
+	driveStatus := 1
```
[the remark this answers is not in the document]
Comment (r-scheele) [reply pkg/device/sysfs_linux.go:117]: `driveStatus` is used to track the status of the drive, so it can also be exported as a metric. Should i still remove it?

**Comment by kannappanr** [comment]
@balamurugana Can you PTAL

File: pkg/device/sysfs_linux.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply pkg/device/sysfs_linux.go:117]: @r-scheele ^^

**Comment by balamurugana** [comment]
@r-scheele I sent a PR https://github.com/r-scheele/directpv/pull/1 to your branch in order to avoid many review cycles. Please check and merge it accordingly.

**Comment by Praveenrajmani** [comment]
closing this as https://github.com/minio/directpv/pull/955 PR got merged
